docs/build-provenance-for-all-registries.md

```diff
@@ -0,0 +1,104 @@
+
+Next you’ll construct the build provenance attestations. We recommend using [in-toto attestations](https://github.com/in-toto/attestation/blob/main/spec/README.md#in-toto-attestation-framework-spec) for formatting your attestation document, and [SLSA provenance](https://slsa.dev/provenance/v1) for the predicates in the in-toto attestation document. See [npm CLI's provenance.js](https://github.com/npm/cli/blob/latest/workspaces/libnpmpublish/lib/provenance.js) for a reference implementation.
+
+Now you’re ready to sign the package! You’ll create an ephemeral key pair, and then send to Fulcio the OIDC token, your ephemeral public key, the OIDC subject signed with your private key (the “challenge value”), and your build artifact signed with your private key. See [sigstore-js' sign.ts](https://github.com/sigstore/sigstore-js/blob/main/packages/client/src/sign.ts#L110) for a reference implementation.
+
+Fulcio will return to you a X.509 signing certificate, with the build provenance properties (including your public key) as OIDs in that certificate. You can now throw away your ephemeral private key.
+
+But we aren’t quite ready to send things to the registry - first you’ll sign your build attestation document (with the Fulcio signing certificate), publish it to Sigstore’s Rekor to get a transparency log entry, and use that to create your Sigstore signature bundle with your signatures over your build artifact, the build attestation document, your transparency log entry, and a timestamp.
+
+You might be wondering how much of this Sigstore interaction you need to implement yourself. Ideally there would be a mature Sigstore library in a language supported by your ecosystem tooling. There are currently libraries for [JavaScript](https://github.com/sigstore/sigstore-js), [Go](https://github.com/sigstore/sigstore-go), [Python](https://github.com/sigstore/sigstore-python), as well libraries under development for [Rust](https://github.com/sigstore/sigstore-rs), [Ruby](https://github.com/sigstore/ruby-sigstore), and [Java](https://github.com/sigstore/sigstore-java). See [sigstore-conformance](https://github.com/sigstore/sigstore-conformance) to ensure client libraries are behaving as expected.
+
+
+# Extending the registry to support build provenance
+
+Now we’re ready to publish the package to the registry, but we need a way to authorize the request. Instead of requiring maintainers to store their password or an unscoped API token in the cloud CI/CD provider, we recommend allowing maintainers to set up a trust relationship between the cloud CI/CD system and your registry via OIDC. For an example of this, see <https://docs.pypi.org/trusted-publishers/>. In fact, registries can implement OIDC trust relationships to authorize publish requests without needing any support from Sigstore client libraries. So this is an excellent scoped project to start with, which is exactly what PyPI did.
```

Comment:
Wait, why do I need to do this? The build provenance already includes proof of the cloud CI/CD OIDC dance. If either of them are invalid, or worse they are both valid but don't match, something is very wrong. If the bill provenance is valid, then the cloud CI/CD OIDC is redundant. The general adage is to make invalid states on representable. What is the justification here for the redundant or invalid inputs?

Reply:
You certainly could use the Fulcio signing certificate to also authorize the package publish!

This suggestion is more around scoping and ordering feature delivery. PyPI's Trusted Publisher feature delivered a security improvement much more quickly than npm's build provenance feature. You could tie these two features together in one big release, but you don't have to.

Reply:
Not just this paragraph, this hole draft seems odd. The way you describe it in the discussion makes it sound `descriptive`. Like it is just what npm did and how to interact with that implementation. But the text itself comes across to me as `proscriptive`. Like "we recommend", and how it does not discuss the alternatives that were considered. Who is your audience and what are you trying to tell them? It is reasonable to not want to write the definitive treatise that describes all of the reasonable possibilities and why each one should be considered, in which case the text could just be explicit about when it is describing npm's choices.

Reply:
The goal here is for the WG to produce something proscriptive, for other software repositories, based on the prior experience of npm/PyPI.

Reply:
I've reworded this section to take this feedback into account!